### PR TITLE
feat: add support of CRNN in Pytorch

### PR DESCRIPTION
**Pull request by charlesmindee**
This PR adds the support for CRNN in Pytorch. The model is implemented with VGG16 backbone implemented in backbones/vgg/pytorch.py. The model (including loss and postprocessor to decode logits) are implemented in recognition/crnn/pytorch.py.

Added torch unitests for crnn.

Any feedback is welcome!

### CODE REVIEW

**Comment by charlesmindee** [comment]
Moved the initialization of max_length to core model, created a resnet folder in backbones with a tensorflow implementation only.

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Some merging is required but the rest looks fine!

File: doctr/models/recognition/crnn/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    ) -> List[Tuple[str, float]]:
+        """Implements best path decoding as shown by Graves (Dissertation, p63), highly inspired from
+        <https://github.com/githubharald/CTCDecoder>`_.
+
+        Args:
+            logits: model output, shape: N x C x T
+            vocab: vocabulary to use
+            blank: index of blank label
+
+        Return:
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:52]: Returns:

File: doctr/models/recognition/crnn/pytorch.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:53]: Missing the return description

File: doctr/models/recognition/crnn/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    def __call__(
+        self,
+        logits: torch.Tensor
+    ) -> List[Tuple[str, float]]:
+        """
+        Performs decoding of raw output with CTC and decoding of CTC predictions
+        with label_to_idx mapping dictionnary
+
+        Args:
+            logits: raw output of the model, shape BATCH_SIZE X NUM_CLASSES + 1 X SEQ_LEN
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:81]: `shape (N, C + 1, seq_len)` (the `X` or `*` is confusing otherwise)

File: doctr/models/recognition/crnn/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    Sequence Recognition and Its Application to Scene Text Recognition" <https://arxiv.org/pdf/1507.05717.pdf>`_.
+
+    Args:
+        feature_extractor: the backbone serving as feature extractor
+        vocab: vocabulary used for encoding
+        rnn_units: number of units in the LSTM layers
+        cfg: configuration dictionary
+    """
+
+    _children_names: List[str] = ['feat_extractor', 'decoder', 'postprocessor']
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:102]:
```python
_children_names: List[str] = ['feat_extractor', 'decoder', 'linear', 'postprocessor']
```

otherwise the repr won't show the FC layers

File: doctr/models/recognition/crnn/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            The loss of the model on the batch
+        """
+        gt, seq_len = self.compute_target(target)
+        batch_len = model_output.shape[0]
+        input_length = model_output.shape[1] * torch.ones(size=(batch_len,), dtype=torch.int32)
+        # N x T x C -> T x N x C
+        logits = model_output.permute(1, 0, 2)
+        probs = F.log_softmax(logits, dim=-1)
+        ctc_loss_fn = nn.CTCLoss(blank=len(self.vocab))
+        ctc_loss = ctc_loss_fn(
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:145]:
```python
ctc_loss = F.ctc_loss(
    probs, torch.from_numpy(gt), input_length, torch.tensor(seq_len, dtype=torch.int), len(self.vocab)
)
```
if you want to do everything in the call

otherwise instantiate the ctcloss module in the constructor ;)

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Small changes remaining and we're good to merge!

File: doctr/models/recognition/crnn/pytorch.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+        self.decoder = nn.LSTM(
+            input_size=1 * 512, hidden_size=rnn_units, batch_first=True, num_layers=2, bidirectional=True
+        )
+
+        # features units = 2 * rnn_units because bidirectional layers
+        self.linear = nn.Linear(in_features=2 * rnn_units, out_features=len(vocab) + 1)
+
+        self.postprocessor = CTCPostProcessor(vocab=vocab)
+
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:122]: Oh yeah I forgot: add some layer initialization here:
```python
for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight.data, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    m.bias.data.zero_()
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1.0)
                m.bias.data.zero_()
```

File: doctr/models/recognition/crnn/pytorch.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
+# Copyright (C) 2021, Mindee.
+
+# This program is licensed under the Apache License version 2.
+# See LICENSE or go to <https://www.apache.org/licenses/LICENSE-2.0.txt> for full license details.
+
+from copy import deepcopy
+from itertools import groupby
+import numpy as np
```
Comment (fg-mindee) [inline doctr/models/recognition/crnn/pytorch.py:8]: this import isn't being used

File: test/pytorch/test_models_recognition.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import pytest
+import numpy as np
```
Comment (fg-mindee) [inline test/pytorch/test_models_recognition.py:2]: you can remove this import

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Small imports to  be removed and we can merge!

**Review by fg-mindee — APPROVED** [review]
Thanks for the edits!
